pysd/py_backend/output.py

```diff
@@ -0,0 +1,516 @@
+                    elif isinstance(comp_vals, np.ndarray):
+                        if comp_vals.size == 1:
+                            self.ds[key][:] = comp_vals
+                        else:
+                            self.ds[key][:] = comp_vals
+                    else:
+                        self.ds[key][:] = comp_vals
+                except ValueError:
+                    warnings.warn(
+                        f"The dimensions of {key} in the results "
+                        "do not match the declared dimensions for this "
+                        "variable. The resulting values will not be "
+                        "included in the results file.")
+
+        self.step += 1
+
+    def postprocess(self, **kwargs):
+        """
+        Closes netCDF4 Dataset.
+
+        Returns
+        -------
+        None
+        """
+
+        # close Dataset
+        self.ds.close()
+
+        print(f"Results stored in {self.out_file}")
+
+    def add_run_elements(self, model, capture_elements):
+        """
+        Adds constant elements to netCDF4 Dataset.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: list
+            List of constant elements
+
+        Returns
+        -------
+        None
+        """
+        # creating variables in capture_elements
+        # TODO we are looping through all capture elements twice. This
+        # could be avoided
+        self.__create_ds_vars(model, capture_elements, add_time=False)
+
+        self.update(model, capture_elements)
+
+    def __create_ds_vars(self, model, capture_elements, add_time=True):
+        """
+        Create new variables in a netCDF4 Dataset from the capture_elements.
+
+        Parameters
+        ----------
+        model: pysd.Model
+            PySD Model object
+        capture_elements: set
+            Which model elements to capture - uses pysafe names.
+        add_time: bool
+            Whether to add a time as the first dimension for the variables.
+
+        Returns
+        -------
+        None
+
+        """
+
+        for key in capture_elements:
+            comp = getattr(model.components, key)
+            comp_vals = comp()
+
+            dims = ()
+
+            if isinstance(comp_vals, (xr.DataArray, np.ndarray)):
+                if comp.subscripts:
+                    dims = tuple(comp.subscripts)
+
+            if add_time:
+                dims = ("time",) + dims
+
+            self.ds.createVariable(key, "f8", dims, compression="zlib")
+
+            # adding units and description as metadata for each var
+            self.ds[key].units = model.doc.loc[
+                        model.doc["Py Name"] == key,
+                        "Units"].values[0] or "Missing"
+            self.ds[key].description = model.doc.loc[
+                        model.doc["Py Name"] == key,
+                        "Comment"].values[0] or "Missing"
```

Comment:
I would save all the information in the docs. Including original name, type, subtype, limits...

Reply:
you mean you wouldn't add it to the Dataset?

Reply:
or you mean that you'd add them all in the Dataset?

Reply:
I mean add them all in the Dataset

Reply:
Done!